### PR TITLE
Build: add a new "Cancelled" final state

### PR DESCRIPTION
**Pull request by humitos**
This commit introduces a new build final state: "Cancelled".
This state is shown in the build listing page and also in the build details
page. The build will end up on this state when:

- The user cancels the build by clicking on the button
- The build is cancelled due to inactivity
- The build is cancelled because it was duplicated

Ther are a lot of changes required for this:

 - HTML template logic for build details and listing pages
 - API changes to add a new value
 - Knockout.js logic for build details page
 - `Build` model changes to support a new state
 - Update different places of core logic to change checking for `finished` state
   to `BUILD_FINAL_STATES` instead
 - Documentation changes to mention the new state
 - Celery build handler `after_return` to decide if it's `finished` or
   `cancelled` the ending state
 - Logic to decide if the build should be killed due to inactivity

### Build ending up in a cancelled state

Screenshot_2022-05-04_13-13-15

### Build cancelled by the user immediately

Screenshot_2022-05-04_13-13-42

### Build cancelled shown on listing page

Screenshot_2022-05-04_13-13-57

Reference https://github.com/readthedocs/readthedocs.org/issues/9100

### CODE REVIEW

**Comment by ericholscher** [comment]
I think overall this makes sense. I feel like @agjohnson had a bit more context in the discussion thread, so happy to let him take primary reviewer here. Happy to do a technical review if we're 👍 on the direction, if that's useful.

File: readthedocs/templates/builds/build_detail.html [hunk 1]
```diff
@@ -138,9 +138,14 @@
         {% trans "Build completed" %}
       </span>
       <span class="build-state-failed"
-            data-bind="visible: finished() && !success()"
+            data-bind="visible: finished() && !success() && !cancelled()"
```
Comment (agjohnson) [inline readthedocs/templates/builds/build_detail.html:141]: This is the type of logic that I'll move to strictly JS and would like out of templates eventually. For now this is fine though.

File: readthedocs/templates/projects/project_dashboard_base.html [hunk 2]
```diff
@@ -77,6 +77,7 @@ <h3>{% trans "Projects" %}</h3>
                               <span class="build-state build-state-passing">{% trans "passing" %}</span>
                             {% else %}
                               <span class="build-state build-state-failing">{% trans "failing" %}</span>
+                              {# TODO: What whould be show here when "not build.sucess and cancelled"? #}
```
Comment (agjohnson) [inline readthedocs/templates/projects/project_dashboard_base.html:80]: I'm getting rid of this concept in the new templates. There is no singular status that makes sense here in any regard. That is, we are showing the status of the project based on the `get_latest_build` -- a failure on some minor version can make the entire project look failing.

Instead, the new templates don't show this status at all, but show individual version statuses instead. Even in that case, I don't feel we show the version as "cancelled" though.

**Review by agjohnson — APPROVED** [review]
This does highlight why I want to centralize this logic. I don't have a great grasp on what could have been missed here, but this looks thorough enough to test.

**Pull request merged**